### ISSUE TITLE
Unable to push to a new branch on "upstream" remote

### ISSUE THREAD

**Issue by kaushalmodi** [magit/magit#3626]
- I created a new local branch "go-env-for-tests".
- `P p` worked fine, pushed to "origin/go-env-for-tests"
- But when I tried to configure `P u` to push to "upstream/go-env-for-tests", I got:

```
Debugger entered--Lisp error: (wrong-type-argument listp "upstream/go-env-for-tests")
  magit-set-upstream-branch("go-env-for-tests" "upstream/go-env-for-tests")
  magit-push-current-to-upstream(nil "upstream/go-env-for-tests")
  funcall-interactively(magit-push-current-to-upstream nil "upstream/go-env-for-tests")
  call-interactively(magit-push-current-to-upstream)
  magit-invoke-popup-action(117)
  funcall-interactively(magit-invoke-popup-action 117)
  call-interactively(magit-invoke-popup-action nil nil)
  command-execute(magit-invoke-popup-action)
``` 

*(I got the same error on running the command generated using `M-x magit-emacs-Q-command`.)*

>  M-x magit-version RET

Magit 20181022.1221, Git 2.18.0.547.g1d89318c4, Emacs 27.0.50, gnu/linux

**Comment by hugot** [comment]
I encountered the same issue, it is seems to have been caused by the addition of a new function for this purpose: ddda6088136f72d811bf9a073f9bf796f7f48f89 . This commit made it so that `magit-set-branch*merge/remote` calls the new function `magit-set-upstream-branch`. `magit-set-upstream-branch` expects to recieve a cons cell, while the function itself used to accept strings, which is why  `magit-read-upstream-branch` returns a string.

The commit mentioned above also caused `magit-branch-reset` to no longer work in cases where it would need to set the upstream of the resetted branch. Executing
```elisp
(with-current-buffer "magit: dotfiles" (magit-branch-reset "master" "origin/random" nil t))
```
confirmed that for me. I think this is because `magit-branch-reset` used `magit-set-branch*merge/remote` before, which used to accept a string as branch name.
